src/main/java/org/prebid/server/bidder/connectAd/ConnectAdBidder.java

```diff
@@ -0,0 +1,190 @@
+            } catch (PreBidException e) {
+                errors.add(BidderError.badInput(e.getMessage()));
+            }
+        }
+        if (CollectionUtils.isNotEmpty(errors)) {
+            errors.add(BidderError.badInput("Error in preprocess of Imp"));
+            return Result.of(Collections.emptyList(), errors);
+        }
+        final BidRequest outgoingRequest = request.toBuilder().imp(processedImps).build();
+
+        final String body = mapper.encode(outgoingRequest);
+
+        final MultiMap headers = HttpUtil.headers();
+        updateHeadersDueToRequest(headers, request);
+
+        return Result.of(Collections.singletonList(
+                HttpRequest.<BidRequest>builder()
+                        .method(HttpMethod.POST)
+                        .uri(endpointUrl)
+                        .headers(headers)
+                        .payload(outgoingRequest)
+                        .body(body)
+                        .build()),
+                errors);
+    }
+
+    private ExtImpConnectAd parseImpExt(Imp imp) {
+        final ExtImpConnectAd extImpConnectAd;
+        try {
+            extImpConnectAd = mapper.mapper().convertValue(imp.getExt(), CONNECTAD_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(String.format("Impression id=%s, has invalid Ext", imp.getId()));
+        }
+
+        if (extImpConnectAd.getSiteId().equals(NumberUtils.INTEGER_ZERO)) {
+            throw new PreBidException(String.format("Impression id=%s, has no siteId present", imp.getId()));
+        }
+        return extImpConnectAd;
+    }
+
+    private Imp updateImp(Imp imp, Integer secure, ExtImpConnectAd extImp) {
+
+        final Imp.ImpBuilder updatedImp = imp.toBuilder().tagid(extImp.getSiteId().toString()).secure(secure);
+        if (extImp.getBidfloor().compareTo(BigDecimal.ZERO) != NumberUtils.INTEGER_ZERO) {
+            updatedImp.bidfloor(extImp.getBidfloor()).bidfloorcur("USD");
+        }
+
+        final Banner banner = imp.getBanner();
+        if (banner == null) {
+            throw new PreBidException("We need a Banner Object in the request");
+        }
+
+        if (banner.getW() == null && banner.getH() == null) {
+            if (CollectionUtils.isEmpty(banner.getFormat())) {
+                throw new PreBidException("At least one size is required");
+            }
+            final Format format = banner.getFormat().get(0);
+            final List<Format> slicedFormatList = banner.getFormat();
+
+            slicedFormatList.remove(0);
+            updatedImp.banner(banner.toBuilder().format(slicedFormatList).w(format.getW()).h(format.getH()).build());
+        }
+
+        return updatedImp.build();
+    }
+
+    private void updateHeadersDueToRequest(MultiMap headers, BidRequest request) {
+        final Device device = request.getDevice();
+
+        if (device != null) {
+            headers.add("User-Agent", device.getUa());
```

Comment:
It may be better to create a new method for adding headers here which will check values for null/empty.